flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceReader.java

```diff
@@ -231,16 +236,7 @@ private void setCurrentReader(int index) {
         reader.start();
         currentSourceIndex = index;
         currentReader = reader;
-        currentReader
-                .isAvailable()
-                .whenComplete(
-                        (result, ex) -> {
-                            if (ex == null) {
-                                availabilityFuture.complete(result);
-                            } else {
-                                availabilityFuture.completeExceptionally(ex);
-                            }
-                        });
```

Comment:
Note that this works also in case of switch events because the reader will be closed then which will complete its availability future.

Reply:
Thanks for taking a look!

The problem isn't only that the future is not renewed after each completion. It also because `whenComplete()` executes, the future is marked at complete and is never renewed (for the underlying reader to tell to the runtime to stop invoking pollNext()).

I think it can be refactored in other means. Proxying the `currentReader.isAvailable()` could also work since the underlying reader itself handles renewing the availability future. I'll take a look

Reply:
Why wouldn't the above solution work (alongside with the other proposed changes)? The future would always be renewed once completed. So the runtime would stop polling for any new calls of `isAvailable()`.

Reply:
If I understand the process in the alternative solution:
1. A new reader is set.
2. When the reader future is done, renew the future.

What is the process that renews the renewed future in step 2? Such a process would need to be recursively ran outside of this reader switch handling method. And repeatedly signaling completeness from the underlying reader and renewing again and again. (This is why PR introduces this additional complexity)

Also in the code snippet above, I think there could be a need for synchronization on the future, since `whenComplete()` would not be guaranteed to run on the task thread, unless we use `whenCompleteAsync()` and pass the task executor, which we don't have access to.

Reply:
Makes sense. We only really deal with one future here at a time, so I agree that the helper class adds more complexity than necessary. I worked on integrating your idea and enhanced the unit test. 

However, I did not include lines 98-102 (https://github.com/apache/flink/compare/master...mxm:flink:FLINK-27479?expand=1#diff-2a9fc4178aac9e00bad9d172c1011d50a2ecd17ada6ca6260db40728bb978723R98-R102). The underlying reader should not complete it's underlying future if it is reporting 'END_OF_INPUT'